### PR TITLE
[RFR]Added crud test for non global filters

### PR DESCRIPTION
**Pull request by mmojzis**
Extending test_advanced_search parametrized tests.
Added filter crud test for non global filters:
1) Creates filter
2) Checks whether filter was created
3) Changes first expression element value
4) Checks whether first expression element value was changed
5) Deletes filter
6) Checks whether My Filters Accordion disappeared or filter was deleted

There are still many things which are not working correctly:
1) I need uncollectif specifically for this test, but I also need global uncollectif working - probably add to global uncollectif to uncollect only when this test is executed.
2) There is change in naming for Job Templates in 5.10. VersionPick didn't work with changing these parameters so I had to add parameters for both 5.9 and 5.10 and then uncollect them depending on appliance version. I don't think this approach is correct as it creates two parametrized tests for one parameter...
3) There are three types of possible tree views with accordion:
- One mostly used - My filters as accordion and my_filters class used for navigation
- Accordion name is different same as class used for navigation(tree is used for navigation instead) - that's reason why I had to add fifth param tuple, which has class name and also category path name.
- Last is only for My Services where it's implemented differently - Services as Accordion and tree is used for navigation. This is why I had to check whether fifth param is list or string.)
I think this could be solved better but I had no idea how to do it.
Also I thank you in advance for comments and enhancements.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 784dc4421c4caf014eab9337854decd0e8806ce0:

`cfme/ansible_tower/explorer.py`:
- :red_circle: Line 27:26: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 27:39: **F821** *undefined name 'Version'*
- :red_circle: Line 88:32: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 88:45: **F821** *undefined name 'Version'*
- :red_circle: Line 157:57: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 157:70: **F821** *undefined name 'Version'*
- :bangbang: Line 157:101: **E501** *line too long (121 > 100 characters)*

`cfme/cloud/tenant.py`:
- :bangbang: Line 18:101: **E501** *line too long (102 > 100 characters)*

`cfme/infrastructure/config_management.py`:
- :bangbang: Line 151:5: **E303** *too many blank lines (2)*

`cfme/tests/webui/test_advanced_search.py`:
- :red_circle: Line 11:1: **F401** *'cfme.utils.blockers.BZ' imported but unused*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit dc57e323c7c8b39f6a28f73ee995161aafb5af39:

`cfme/ansible_tower/explorer.py`:
- :red_circle: Line 27:26: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 27:39: **F821** *undefined name 'Version'*
- :red_circle: Line 88:32: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 88:45: **F821** *undefined name 'Version'*
- :red_circle: Line 157:57: **F821** *undefined name 'VersionPick'*
- :red_circle: Line 157:70: **F821** *undefined name 'Version'*
- :bangbang: Line 157:101: **E501** *line too long (121 > 100 characters)*

`cfme/cloud/tenant.py`:
- :bangbang: Line 18:101: **E501** *line too long (102 > 100 characters)*

`cfme/infrastructure/config_management.py`:
- :bangbang: Line 151:5: **E303** *too many blank lines (2)*

`cfme/tests/webui/test_advanced_search.py`:
- :red_circle: Line 11:1: **F401** *'cfme.utils.blockers.BZ' imported but unused*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9cea7c40ef08fd9e84f62f64ff2b2f0fe143f4e5:

`cfme/tests/webui/test_advanced_search.py`:
- :red_circle: Line 11:1: **F401** *'cfme.utils.blockers.BZ' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/ansible_tower/explorer.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from widgetastic.widget import Text, View
 from widgetastic_patternfly import Dropdown
 from widgetastic_manageiq import Search, ItemsToolBarViewSelector, Button, Accordion, ManageIQTree
+from widgetastic.utils import Version, VersionPick
```
Comment (quarckster) [inline cfme/ansible_tower/explorer.py]: **REQUIRED**
```python
from cfme.utils.version import Version, VersionPicker
```

File: cfme/ansible_tower/explorer.py [hunk 2]
```diff
@@ -85,7 +86,8 @@ class TowerExplorerJobTemplatesAllView(TowerExplorerView):
     def is_displayed(self):
         return (
             self.in_tower_explorer and
-            self.title.text == 'All Ansible Tower Job Templates' and
+            self.title.text == VersionPick({Version.lowest(): 'All Ansible Tower Job Templates',
+                                       '5.10': 'All Ansible Tower Templates'}) and
```
Comment (quarckster) [inline cfme/ansible_tower/explorer.py]: **REQUIRED**
```python
self.title.text == VersionPicker({
    Version.lowest(): 'All Ansible Tower Job Templates',
    '5.10': 'All Ansible Tower Templates'}).pick(self.appliance.version)
```

File: cfme/ansible_tower/explorer.py [hunk 3]
```diff
@@ -153,4 +155,5 @@ class TowerExplorerJobTemplatesAll(CFMENavigateStep):
     prerequisite = NavigateToAttribute('appliance.server', 'AnsibleTowerExplorer')
 
     def step(self, *args, **kwargs):
-        self.view.sidebar.job_templates.tree.click_path('All Ansible Tower Job Templates')
+        self.view.sidebar.job_templates.tree.click_path(VersionPick({Version.lowest():
+            'All Ansible Tower Job Templates', '5.10': 'All Ansible Tower Templates'}))
```
Comment (quarckster) [inline cfme/ansible_tower/explorer.py]: ^

File: cfme/cloud/availability_zone.py [hunk 4]
```diff
@@ -81,6 +81,13 @@ def is_displayed(self):
     toolbar = View.nested(AvailabilityZoneToolBar)
     including_entities = View.include(AvailabilityZoneEntities, use_parent=True)
 
+    @View.nested
+    class my_filters(Accordion):  # noqa
+        ACCORDION_NAME = "My Filters"
+
+        navigation = BootstrapNav('.//div/ul')
+        tree = ManageIQTree
```
Comment (quarckster) [inline cfme/cloud/availability_zone.py]: **REQUIRED**
```python
tree = ManageIQTree()
```

File: cfme/cloud/flavor.py [hunk 5]
```diff
@@ -69,6 +69,13 @@ class FlavorAllView(FlavorView):
     search = View.nested(Search)
     including_entities = View.include(FlavorEntities, use_parent=True)
 
+    @View.nested
+    class my_filters(Accordion):  # noqa
+        ACCORDION_NAME = "My Filters"
+
+        navigation = BootstrapNav('.//div/ul')
+        tree = ManageIQTree
```
Comment (quarckster) [inline cfme/cloud/flavor.py]: ^

File: cfme/cloud/host_aggregates.py [hunk 6]
```diff
@@ -27,6 +27,13 @@ def in_host_aggregates(self):
         return (self.logged_in_as_current_user and
                 self.navigation.currently_selected == ['Compute', 'Clouds', 'Host Aggregates'])
 
+    @View.nested
+    class my_filters(Accordion):  # noqa
+        ACCORDION_NAME = "My Filters"
+
+        navigation = BootstrapNav('.//div/ul')
+        tree = ManageIQTree
```
Comment (quarckster) [inline cfme/cloud/host_aggregates.py]: ^

File: cfme/cloud/tenant.py [hunk 7]
```diff
@@ -91,6 +92,13 @@ class TenantAllView(TenantView):
     paginator = PaginationPane()
     search = View.nested(Search)
 
+    @View.nested
+    class my_filters(Accordion):  # noqa
+        ACCORDION_NAME = "My Filters"
+
+        navigation = BootstrapNav('.//div/ul')
+        tree = ManageIQTree(tree_id='')
```
Comment (quarckster) [inline cfme/cloud/tenant.py]: **REQUIRED**
```python
tree = ManageIQTree(tree_id='')
```

File: cfme/common/physical_server_views.py [hunk 8]
```diff
@@ -2,13 +2,13 @@
...
-from widgetastic_manageiq import (
-    BaseEntitiesView, JSBaseEntity, BaseListEntity, BaseQuadIconEntity, BaseTileIconEntity,
-    BootstrapTreeview, Button, ItemsToolBarViewSelector, Search, SummaryTable, TimelinesView,
-    BaseNonInteractiveEntitiesView, ManageIQTree, NonJSBaseEntity)
-from widgetastic_patternfly import BreadCrumb, Dropdown, Accordion
-
 from cfme.base.login import BaseLoggedInPage
+from widgetastic_patternfly import (
+    BootstrapNav, BreadCrumb, Dropdown, Accordion)
+from widgetastic_manageiq import (
```
Comment (quarckster) [inline cfme/common/physical_server_views.py]: **REQUIRED**
These imports were in a correct place, please revert it back.

File: cfme/tests/webui/test_advanced_search.py [hunk 9]
```diff
@@ -111,3 +135,86 @@ def test_can_open_advanced_search(params, appliance):
         pytest.fail("Advanced search cannot be opened for {}_{}".format(
             params[2], params[1].lower())
         )
+
+
+# @pytest.mark.uncollectif(lambda params: ((params[0] == TemplatesImages and BZ(1626579)) or
+#                                          (params[0] == MyService and BZ(1627078))))
+def test_filter_crud(params, appliance):
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
It's a mess, this function is too big, consider to divide it on some smaller functions.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/tests/webui/test_advanced_search.py (hunk 9 above)
Comment (mmojzis) [reply cfme/tests/webui/test_advanced_search.py]: Could you help me how can I divide it in multiple smaller functions? Everything I tried broke the uncollectif.

**Comment by dajoRH** [comment]
Lint report for commit 2e9d36a145d84cecd0540577ed96557c8b1fc307:

`cfme/ansible_tower/explorer.py`:
- :bangbang: Line 160:101: **E501** *line too long (116 > 100 characters)*

`cfme/common/physical_server_views.py`:
- :bangbang: Line 15:1: **E303** *too many blank lines (3)*

`cfme/common/provider_views.py`:
- :bangbang: Line 47:1: **E305** *expected 2 blank lines after class or function definition, found 0*
- :bangbang: Line 47:2: **E999** *SyntaxError: invalid syntax*
- :bangbang: Line 47:7: **E225** *missing whitespace around operator*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7bf3d813bc2c91bee958e0dc777fbf7376f5170e:

`cfme/common/provider_views.py`:
- :red_circle: Line 5:1: **F401** *'widgetastic.widget.ParametrizedView' imported but unused*
- :red_circle: Line 313:16: **F821** *undefined name 'ManageIQTree'*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/ansible_tower/jobs.py [hunk 10]
```diff
@@ -26,6 +26,13 @@ def in_jobs(self):
         return (self.logged_in_as_current_user and
                 self.navigation.currently_selected == ['Automation', 'Ansible Tower', 'Jobs'])
 
+    @View.nested
+    class my_filters(Accordion):  # noqa
```
Comment (mshriver) [inline cfme/ansible_tower/jobs.py]: How about defining this standard view in `widgetastic_manageiq` and using `View.nested(FilterAccordion)` in all of the places where its defined in this PR?

File: cfme/ansible_tower/explorer.py (hunk 2 above)
Comment (mmojzis) [reply cfme/ansible_tower/explorer.py]: Added versionpicker: https://github.com/ManageIQ/integration_tests/pull/7785/files#diff-ca2caa6aac1e64d4303659fee714d350R89 But it fails on that it hasn't self.appliance.

**Comment by mmojzis** [comment]
@quarckster All these requested changes are in the code now.

File: cfme/tests/webui/test_advanced_search.py [hunk 11]
```diff
@@ -1,127 +1,237 @@
...
+    view = _navigation(param, appliance)
+    # create
+    view.search.save_filter(
+        "fill_field({}, =, {})".format(param.filter, filter_value), filter_name)
+    view.search.close_advanced_search()
+    view.flash.assert_no_error()
+    # read
+    if param.my_filters:
+        if isinstance(param.my_filters, tuple):
+            filters = operator.attrgetter(param.my_filters[0])(view)
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
I think this is slightly more readable.
```python
filters = getattr(view, param.my_filters[0])
```

File: cfme/tests/webui/test_advanced_search.py [hunk 12]
```diff
@@ -1,127 +1,237 @@
...
+            _filter_displayed(filters, filters.tree.has_path(param.my_filters[1], "My Filters",
+                                                             filter_name))
+        else:
+            filters = operator.attrgetter(param.my_filters)(view)
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
```python
filters = getattr(view, param.my_filters)
```

File: cfme/tests/webui/test_advanced_search.py [hunk 13]
```diff
@@ -1,127 +1,237 @@
...
+        _filter_displayed(filters, filters.navigation.has_item(filter_name))
+    # update
+    _select_filter(filters, filter_name, param)
+    view.search.open_advanced_search()
+    view.search.advanced_search_form.search_exp_editor.select_first_expression()
+    view.search.advanced_search_form.search_exp_editor.fill_field(field=param.filter,
+                                                                  key='=',
+                                                                  value=filter_value_updated)
+    view.search.advanced_search_form.save_filter_button.click()
+    view.search.advanced_search_form.save_filter_button.click()
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQURIED**
Please add a comment why this line is required twice.

File: cfme/tests/webui/test_advanced_search.py [hunk 14]
```diff
@@ -1,127 +1,237 @@
...
+    view.search.close_advanced_search()
+    _select_filter(filters, filter_name, param)
+    # read after update
+    view.search.open_advanced_search()
+    exp_text = view.search.advanced_search_form.search_exp_editor.expression_text
+    if filter_value_updated not in exp_text:
+        pytest.fail("Filter wasn't changed!")
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
```python
assert filter_value_updated in exp_text, "Filter wasn't changed!"
```

File: cfme/tests/webui/test_advanced_search.py [hunk 15]
```diff
@@ -1,127 +1,237 @@
...
+    # delete
+    view.search.delete_filter()
+    view.search.close_advanced_search()
+    if param.my_filters:
+        if filters.is_displayed:
+            if isinstance(param.my_filters, tuple):
+                if filters.tree.has_path(param.my_filters[1], "My Filters", filter_name):
+                    pytest.fail("Filter wasn't deleted!")
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
```python
assert not filters.tree.has_path(param.my_filters[1], "My Filters", filter_name), "Filter wasn't deleted!"
```

File: cfme/tests/webui/test_advanced_search.py [hunk 16]
```diff
@@ -1,127 +1,237 @@
...
+            else:
+                if filters.tree.has_path("My Filters", filter_name):
+                    pytest.fail("Filter wasn't deleted!")
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
```python
assert not filters.tree.has_path("My Filters", filter_name), "Filter wasn't deleted!"
```

File: cfme/tests/webui/test_advanced_search.py [hunk 17]
```diff
@@ -1,127 +1,237 @@
...
+    else:
+        if view.my_filters.is_displayed:
+            if view.my_filters.navigation.has_item(filter_name):
+                pytest.fail("Filter wasn't deleted!")
```
Comment (quarckster) [inline cfme/tests/webui/test_advanced_search.py]: **REQUIRED**
```python
assert not view.my_filters.navigation.has_item(filter_name), "Filter wasn't deleted!"
```

File: cfme/ansible_tower/explorer.py (hunk 2 above)
Comment (quarckster) [reply cfme/ansible_tower/explorer.py]: Try
```python
self.title.text == VersionPicker({
    Version.lowest(): 'All Ansible Tower Job Templates',
    '5.10': 'All Ansible Tower Templates'}).pick(self.browser.product_version)

File: cfme/tests/webui/test_advanced_search.py (hunk 11 above)
Comment (mmojzis) [reply cfme/tests/webui/test_advanced_search.py]: @quarckster getattr doesn't work for example with 'sidebar.configured_systems'

File: cfme/ansible_tower/jobs.py (hunk 10 above)
Comment (mmojzis) [reply cfme/ansible_tower/jobs.py]: I can work on refactoring, but merging this PR would save so much time on testing so I prefer to make it merged and then I can play with it and refactor these accordions. Thanks

**Pull request merged**